### PR TITLE
HP-1950/Implement_progressive_prices_API

### PR DESCRIPTION
**Pull request by ValeriyShnurovoy**
## Summary by CodeRabbit

- **New Features**
  - Introduced progressive pricing functionality to handle fees based on overuse of services with defined pricing tiers.
  - Added a class to manage and manipulate progressive price thresholds effectively.

- **Tests**
  - Created unit tests to ensure the accuracy of progressive price calculations and JSON serialization.
  - Added behavioral tests for calculating progressive price fees based on service overuse scenarios.

### CODE REVIEW

File: src/price/ProgressivePrice.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+class ProgressivePrice extends AbstractPrice
+{
+    /* @var ProgressivePriceConditionDto[] $thresholds */
+    protected array $thresholds;
+    /**
+     * @var QuantityInterface prepaid quantity also implies Unit
+     * XXX cannot be null cause Unit is required
+     */
+    protected $prepaid;
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:23]: `QuantityInterface`

File: src/price/ProgressivePrice.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        QuantityInterface $prepaid,
+        array $thresholds,
+        ?PlanInterface $plan = null
+    ) {
+        parent::__construct($id, $type, $target, $plan);
+        $this->prepaid = $prepaid;
+        $this->thresholds = $thresholds;
+    }
+
+    public function getPrepaid()
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:38]: return type

File: src/price/ProgressivePrice.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        return $this->prepaid;
+    }
+
+    public function getThresholds(): array
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:43]: PHPDoc to specity array return type

File: src/price/ProgressivePrice.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        return $this->thresholds;
+    }
+
+    private function prepareThresholds(): void
+    {
+        usort($this->thresholds, function($a, $b)
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:50]: function types

File: src/price/ProgressivePriceThresholdsDto.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+namespace hiqdev\php\billing\price;
+
+use Money\Currency;
+
+class ProgressivePriceThresholdsDto
```
Comment (SilverFire) [inline src/price/ProgressivePriceThresholdsDto.php:9]: DTO is used to pass a big set of options required to build an object in Factory... when you don't want to carry data in arrays.

For ProgressivePriceThreshold it makes more sense to add a named constructor that will create some missing objects, than having DTO + Factory.

File: src/price/ProgressivePrice.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+use hiqdev\php\billing\target\TargetInterface;
+use hiqdev\php\billing\type\TypeInterface;
+use hiqdev\php\units\Quantity;
+use hiqdev\php\units\QuantityInterface;
+use Money\Currency;
+use Money\Money;
+
+class ProgressivePrice extends AbstractPrice
+{
+    /* @var ProgressivePriceConditionDto[] $thresholds */
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:17]: DTO should never be a part of an entity. Use VO instead.

File: src/price/ProgressivePriceThresholdsDto.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+{
+   public float $price;
+
+   public Currency $currency;
+
+   public float $value;
```
Comment (SilverFire) [inline src/price/ProgressivePriceThresholdsDto.php:15]: Depend on **Quantity** object instead of scalar

File: src/price/ProgressivePriceThresholdsDto.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace hiqdev\php\billing\price;
+
+use Money\Currency;
+
+class ProgressivePriceThresholdsDto
+{
+   public float $price;
+
+   public Currency $currency;
```
Comment (SilverFire) [inline src/price/ProgressivePriceThresholdsDto.php:13]: That`s what **Money** is. Use it!

File: tests/behat/ProgressivePrice.feature [hunk 9]
```diff
@@ -0,0 +1,17 @@
+Feature: Progressive price fee
+
+  Background:
+    Given target progressive price for overuse,cdn_traf95_max is     0.0085 USD per Mbps great 0 Mbps
+      And target progressive price for overuse,cdn_traf95_max is     0.0080 USD per Mbps great_equal 500 Mbps
+      And target progressive price for overuse,cdn_traf95_max is     0.0075 USD per Mbps great_equal 600 Mbps
+      And target progressive price for overuse,cdn_traf95_max is     0.0070 USD per Mbps great_equal 700 Mbps
+      And target progressive price for overuse,cdn_traf95_max is     0.0060 USD per Mbps great 800 Mbps
```
Comment (SilverFire) [inline tests/behat/ProgressivePrice.feature:8]: great_equal, great are redundant, right?

File: src/price/ProgressivePriceThresholdsDto.php (hunk 5 above)
Comment (SilverFire) [inline src/price/ProgressivePriceThresholdsDto.php:9]: Why not final?

File: src/price/ProgressivePriceThresholdsDto.php [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+declare(strict_types=1);
+
+namespace hiqdev\php\billing\price;
+
+use Money\Currency;
+
+class ProgressivePriceThresholdsDto
+{
+   public float $price;
```
Comment (SilverFire) [inline src/price/ProgressivePriceThresholdsDto.php:11]: Why public?

File: tests/behat/bootstrap/BillingContext.php [hunk 11]
```diff
@@ -116,6 +118,52 @@ public function enumPrice($type, $price, $currency, $unit, $price2, $currency2,
...
+     */
+    public function createProgressivePrices()
+    {
+        foreach ($this->progressivePrice as $type => $price) {
+            $this->fullPrice([
+                'type' => $type,
+                'price' => 0,
+                'currency' => $price['currency'],
+                'unit' => $price['unit'],
+                'data' => ['thresholds' => $price['thresholds'], 'class' => 'progressive'],
```
Comment (SilverFire) [inline tests/behat/bootstrap/BillingContext.php:162]: Change class

File: tests/unit/price/ProgressivePriceTest.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    private function getThresholds()
+    {
+        return [
+            new ProgressivePriceThresholdsDto(0.0085, new Currency('EUR'), 0),
+            new ProgressivePriceThresholdsDto(0.0080, new Currency('EUR'), 500),
+            new ProgressivePriceThresholdsDto(0.0075, new Currency('EUR'), 600),
+            new ProgressivePriceThresholdsDto(0.0070, new Currency('EUR'), 700),
+            new ProgressivePriceThresholdsDto(0.0060, new Currency('EUR'), 800),
+        ];
+    }
```
Comment (SilverFire) [inline tests/unit/price/ProgressivePriceTest.php:82]: Add a test to cover edge cases. What would happen if I create the following thresholds:

0 -> 6$
100 -> 5$
100 -> 4$
200 -> 3$

File: src/price/ProgressivePrice.php [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        $result = null;
+        $this->prepareThresholds();
+        $usage = $this->calculateUsage($quantity);
+        $quantity = $usage->getQuantity();
+        foreach ($this->thresholds as $key => $threshold) {
+            if  ($threshold->value < $quantity) {
+                if ($key !== count($this->thresholds) - 1) {
+                    $boundary = $quantity - $threshold->value;
+                    $result += $boundary * $threshold->price;
+                    $quantity = $quantity - $boundary;
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:85]: Do not operate scalars. Use proper VOs: **Money**, **Quantity** instead

File: src/price/ProgressivePrice.php [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    public function calculateUsage(QuantityInterface $quantity): ?QuantityInterface
+    {
+        $usage = $quantity->convert($this->prepaid->getUnit())->subtract($this->prepaid);
+
+        if ($usage->isPositive()) {
+            return $usage;
+        }
+
+        return Quantity::create($this->prepaid->getUnit()->getName(), 0);
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:68]: Matches **SinglePrice**.

Why did you prefer to extend AbstractPrice?

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: src/price/ProgressivePrice.php (hunk 14 above)
Comment (ValeriyShnurovoy) [reply src/price/ProgressivePrice.php:68]: I extended the abstract class because it had base methods for plan, target, etc.

File: src/price/ProgressivePrice.php (hunk 13 above)
Comment (ValeriyShnurovoy) [reply src/price/ProgressivePrice.php:85]: i need scalar variable in this code

Comment (SilverFire) [reply src/price/ProgressivePrice.php:85]: Why? `$quantity = $quantity->subtract($boundary)` should work

File: src/price/ProgressivePrice.php (hunk 14 above)
Comment (SilverFire) [reply src/price/ProgressivePrice.php:68]: The same methods are in **SinglePrice**, as it extends **AbstractPrice**. Maybe we can extend it not to duplicate method implementations?

File: src/price/ProgressivePriceThresholdsDto.php (hunk 10 above)
Comment (ValeriyShnurovoy) [reply src/price/ProgressivePriceThresholdsDto.php:11]: Why not? We have a public properties in all DTO class

Comment (SilverFire) [reply src/price/ProgressivePriceThresholdsDto.php:11]: Makes sense. I wrote this question, treating the class as if it is already a VO. Sorry

File: src/price/PriceHelper.php [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        if (!is_numeric($price)) {
+            throw new \InvalidArgumentException('price of the Money must be numeric');
+        }
+        if (is_int($price)) {
+            return new Money($price, new Currency($currency));
+        }
+        if (!is_int($price) && ($price * 100) == ((int) ($price * 100))) {
+            return new Money((int) $price * 100, new Currency($currency));
+        }
```
Comment (SilverFire) [inline src/price/PriceHelper.php:24]: Use `DecimalMoneyParser` instead of multiplying by 100.

File: src/price/ProgressivePrice.php [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    public function calculateUsage(QuantityInterface $quantity): ?QuantityInterface
+    {
+        $usage = $quantity->convert($this->prepaid->getUnit());
+
+        if ($usage->isPositive()) {
+            return $usage;
+        }
+
+        return Quantity::create($this->prepaid->getUnit()->getName(), $quantity->getQuantity());
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:69]: What is a sense of creating a Quantity for a negative value? To giveaway money to customers?

File: src/price/ProgressivePrice.php [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+                    $boundary = $usage->subtract($threshold->quantity());
+                    $result = $result->add(new Money(
+                            (int) $boundary->multiply($threshold->price()->getAmount())->getQuantity(),
+                            $threshold->price()->getCurrency()
+                        )
+                    );
+                    $usage = $usage->subtract($boundary);
+            }
+        }
+        $result = (new DecimalMoneyParser(new ISOCurrencies()))->parse($result->getAmount(), $result->getCurrency());
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:96]: What does this line do? It's the same as `$result = clone $result`

File: src/price/ProgressivePriceThreshold.php [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    public string $unit;
+
+    private function __construct(string $price, string $currency, string $quantity, string $unit)
+    {
+        if ($quantity < 0) {
+            throw new \InvalidArgumentException('Quantity of the progressive price threshold must be positive');
+        }
+
+        $this->price = $price;
+        $this->currency = $currency;
```
Comment (SilverFire) [inline src/price/ProgressivePriceThreshold.php:28]:
```suggestion
        $this->currency = strtoupper($currency);
```

File: src/price/ProgressivePriceThreshold.php [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace hiqdev\php\billing\price;
+
+use hiqdev\php\units\Unit;
+use Money\Money;
+use hiqdev\php\units\Quantity;
+
+class ProgressivePriceThreshold
+{
+    public string $price;
```
Comment (SilverFire) [inline src/price/ProgressivePriceThreshold.php:13]: When I put here `"1"` – it's `0.01 $currency`, or `1 $currency`?

File: src/price/ProgressivePriceThreshold.php [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+            $price->getAmount(),
+            $price->getCurrency()->getCode(),
+            $quantity->getQuantity(),
+            $quantity->getUnit()->getName()
+        );
+    }
+
+    /**
+     * @return Money
+     */
```
Comment (SilverFire) [inline src/price/ProgressivePriceThreshold.php:50]: Excess

File: src/price/ProgressivePriceThreshold.php [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    public function price(): Money
+    {
+        return PriceHelper::buildMoney($this->price, $this->currency);
+    }
+
+    /**
+     * @return Quantity
+     */
```
Comment (SilverFire) [inline src/price/ProgressivePriceThreshold.php:58]: Excess

File: src/price/ProgressivePriceThreshold.php [hunk 22]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare(strict_types=1);
+
+namespace hiqdev\php\billing\price;
+
+use hiqdev\php\units\Unit;
+use Money\Money;
+use hiqdev\php\units\Quantity;
+
```
Comment (SilverFire) [inline src/price/ProgressivePriceThreshold.php:10]: We've put here some specific ideas, such as storing scalars instead of objects. Elaborate this decision in PHPDoc please.

File: src/price/ProgressivePriceThresholds.php [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+namespace hiqdev\php\billing\price;
+
+use hiqdev\php\units\Quantity;
+use InvalidArgumentException;
+use Money\Money;
+
+final class ProgressivePriceThresholds
+{
+    /** @var ProgressivePriceThreshold[] */
+    public array $thresholds;
```
Comment (SilverFire) [inline src/price/ProgressivePriceThresholds.php:14]: Why public?

File: src/price/PriceHelper.php [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+use Money\Currency;
+use Money\Money;
+use Money\Parser\DecimalMoneyParser;
+
+class PriceHelper
+{
+    public static function buildMoney(string $price, string $currency): ?Money
+    {
+        if (!is_numeric($price)) {
+            throw new \InvalidArgumentException('price of the Money must be numeric');
```
Comment (SilverFire) [inline src/price/PriceHelper.php:17]: Use imports

File: src/price/PriceHelper.php [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        if (!self::checkFloat($price)) {
+            return new Money($price, new Currency($currency));
+        }
+        if (!is_int($price) && ($price * 100) == ((int) ($price * 100))) {
+            return (new DecimalMoneyParser(new ISOCurrencies()))->parse($price, new Currency($currency));
```
Comment (SilverFire) [inline src/price/PriceHelper.php:23]: DecimalMoneyParser is a service, don't create services with `new` keyword in a code: use DI

File: src/price/PriceHelper.php [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        return new Money(
+            (int) ($price * self::calculatePriceRate($price)),
+            new Currency($currency)
+        );
+    }
+
+    public static function calculatePriceRate(string $price): int
```
Comment (SilverFire) [inline src/price/PriceHelper.php:31]: `Price Rate` is not a self-descriptive name. It's also misleading.

Rate – is about a ratio. A ratio can be non-integer. I'm thinking about better naming.

Maybe Multiplier? @hiqsol

File: src/price/PriceHelper.php [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        if (self::checkFloat($number)) {
+            return explode('.', $number);
+        }
+        return $number;
+    }
+
+    private static function checkFloat(string $number): bool
+    {
+        return strpos($number, '.') > 0;
```
Comment (SilverFire) [inline src/price/PriceHelper.php:50]: Why not use `is_float()`?

File: src/price/ProgressivePrice.php [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+
+    protected QuantityInterface $prepaid;
+
+    public function __construct(
+        $id,
+        TypeInterface $type,
+        TargetInterface $target,
+        QuantityInterface $prepaid,
+        Money $price,
+        array $thresholds,
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:31]: PHPDoc for array type clarification

File: src/price/ProgressivePrice.php [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+    ) {
+        parent::__construct($id, $type, $target, $plan);
+        $this->thresholds = new ProgressivePriceThresholds($thresholds);
+        $this->price = $price;
+        $this->prepaid = $prepaid;
+    }
+
+    /**
+     * @return ProgressivePriceThresholds
+     */
```
Comment (SilverFire) [inline src/price/ProgressivePrice.php:42]: Excess

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: src/price/ProgressivePriceThreshold.php (hunk 19 above)
Comment (ValeriyShnurovoy) [reply src/price/ProgressivePriceThreshold.php:13]: 1

**Review by SilverFire — APPROVED** [review]
